### DISCUSSION TITLE
Get key events for character keys

### DISCUSSION THREAD

**Discussion by obraunsdorf** [fltk-rs/fltk-rs#905] [Q&A] [answered]
Is it possible with fltk to receive KeyUp events with from character/number keys ('a', 'b,' 1',...) WITHOUT having them assigned to widgets as shortcuts?

**Comment by MoAlyousef** [comment, marked as answer]
Yes. You have app::event_key() and app::event_state().

**Reply by obraunsdorf** [reply]
Thanks for the quick repsonse!
Unfortnuately I don't get it to work. I tried the following code:
```
let event = fltk::app::event();
println!("the following event happend {:?}", event);
let shortcut = fltk::app::event_state();
println!("shortcut: {:?}", shortcut);
```
but it only prints
```
the following event happend Push | KeyDown | KeyUp
shortcut: None
```

From the documentation it seems like the enums fltk::enums::Event and fltk::enums::Shortcut do not contain enum variants that represent characters ('a', 'b', etc.). Is it possible to determine if e.g. the 'a' key was pressed?

**Reply by MoAlyousef** [reply]
So generally you’ll have to listen for events in a handle or callback method, or in the event loop. Also you can use the Display impl (no need for :? to know the event type). As for character ‘a’, you would compare against Key::from_char('a'). 
I’m away from my pc, so I can’t provide an example.

**Reply by MoAlyousef** [reply]
An example:
```rust
use fltk::{prelude::*, enums::*, *};

fn main() {
    let app = app::App::default().with_scheme(app::Scheme::Gtk);

    let mut win = window::Window::default()
        .with_size(800, 600);
    win.end();
    win.show();

    win.handle(|w, ev| match ev {
        Event::KeyUp => {
            let event = app::event();
            println!("the following event happend {}", event);
            let key = app::event_key();
            if let Some(key) = std::char::from_u32(key.bits() as _) {
                println!("the following key was pressed {}", key);
            }
            let shortcut = app::event_state();
            println!("shortcut: {:?}", shortcut);
            true
        },
        _ => false,
    });
    app.run().unwrap();
}
```

**Reply by obraunsdorf** [reply]
Thank you so much for that example! Seeing `key.bits() as u32` as and `Key::from_char() helped a lot. I was able to solve my problem.
I think it was confusing to see `Key` defined as 
```
bitflags! {
    /// Defines the inputted virtual keycode
    pub struct Key: i32 {
        /// None
        const None = 0;
        /// Button
        const Button = 0xfee8;
        /// BackSpace
        const BackSpace = 0xff08;
[...]
```

I didn't recognize the syntax at `Key: i32`. Is this a common pattern with the bitflags! macro?

Anyway, thank you very much for the helpful, quick response! :)

**Reply by MoAlyousef** [reply]
You're welcome :) 
Yes it is common.
